src/KOKKOS/pair_metatensor_kokkos.cpp

```diff
@@ -0,0 +1,663 @@
+                            "found {} CUDA-capable GPUs, but only {} MPI processes on the current node; the remaining GPUs will not be used",
+                            torch::cuda::device_count(), size
+                        );
+                    }
+                }
+
+                // split GPUs between node-local processes using round-robin allocation
+                int gpu_to_use = local_rank % torch::cuda::device_count();
+                available_devices.push_back(torch::Device(torch::kCUDA, gpu_to_use));
+            }
+        } else if (device == "mps") {
+            #if TORCH_VERSION_MAJOR >= 2
+            if (torch::mps::is_available()) {
+                available_devices.push_back(torch::Device("mps"));
+            }
+            #endif
+        } else {
+            error->warning(FLERR,
+                "the model declared support for unknown device '{}', it will be ignored", device
+            );
+        }
+    }
+
+    if (available_devices.empty()) {
+        error->all(FLERR,
+            "failed to find a valid device for the model at '{}': "
+            "the model supports {}, none of these where available",  /// typo: where -> were
+            model_path, torch::str(mts_data->capabilities->supported_devices)
+        );
+    }
+
+    if (requested_device == nullptr) {
+        // no user request, pick the device the model prefers
+        mts_data->device = available_devices[0];
+    } else {
+        bool found_requested_device = false;
+        for (const auto& device: available_devices) {
+            if (device.is_cpu() && strcmp(requested_device, "cpu") == 0) {
+                mts_data->device = device;
+                found_requested_device = true;
+                break;
+            } else if (device.is_cuda() && strcmp(requested_device, "cuda") == 0) {
+                mts_data->device = device;
+                found_requested_device = true;
+                break;
+            } else if (device.is_mps() && strcmp(requested_device, "mps") == 0) {
+                mts_data->device = device;
+                found_requested_device = true;
+                break;
+            }
+        }
+
+        if (!found_requested_device) {
+            error->all(FLERR,
+                "failed to find requested device ({}): it is either "
+                "not supported by this model or not available on this machine",
+                requested_device
+            );
+        }
+    }
+
+    mts_data->model->to(mts_data->device);
+
+    // Handle potential mismatch between Kokkos and model devices
+    if (std::is_same<DeviceType, Kokkos::Cuda>::value) {
+        if (!mts_data->device.is_cuda()) {
+            throw std::runtime_error("Kokkos is running on a GPU, but the model is not on a GPU");
+        }
+    } else {
+        if (!mts_data->device.is_cpu()) {
+            throw std::runtime_error("Kokkos is running on CPU, but the model is not on CPU");
+        }
+    }
```

Comment:
Here for the device check

Reply:
Ah, thanks! 

Should we also change the default device? I.e. if the user is not giving a device, picking the same one kokkos is using instead of the first one in the model capabilities?

Also, this check would fail for alternative kokkos device (HIP, OpenCL, …), so I would rather check explicitly for CPU device instead of defaulting to it.

Reply:
I agree for the second part (fixed in the latest commit).

Regarding the first part, I think users of this interface would be relatively advanced and they would be able to understand the error message and pick the correct device, while allowing us to re-use the device selection code from the other interface. Anyway, if you want to change the device selection code instead of sharing it with the base interface, I have no objections